### ISSUE TITLE
Installation and usage

### ISSUE THREAD

**Issue by DanSchoppe** [realgud/realgud-lldb#1]
I'd like to try realgud-lldb.  I have realgud installed through Melpa.  I'm don't have admin privileges on my OS X machine.  I pulled down the realgud-lldb repo, ran `autogen.sh --prefix=$HOME/.emacs.d/external/realgud-lldb --disable-silent-rules`.  Output:

```
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force -I .
autoreconf: configure.ac: tracing
autoreconf: configure.ac: not using Libtool
autoreconf: running: /Users/me/local/Cellar/autoconf/2.69/bin/autoconf --force
autoreconf: configure.ac: not using Autoheader
autoreconf: running: automake --add-missing --copy --force-missing
configure.ac:4: installing './install-sh'
configure.ac:4: installing './missing'
autoreconf: Leaving directory `.'
Running configure with --enable-maintainer-mode --prefix=/Users/me/.emacs.d/external/realgud-lldb --disable-silent-rules
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... ./install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... yes
checking for emacs... /Users/me/local/bin/emacs
configure: "Checking emacs version"
'compute-lispdir.sh' lispdir install directory override: '/Users/me/local/share/emacs/site-lisp'
checking for emacs... /Users/me/local/bin/emacs
checking where .elc files should go... /Users/me/local/share/emacs/site-lisp
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating lldb/Makefile
config.status: creating test/Makefile
```

I'm not sure if autogen was supposed to actually install realgud or not but I see that it spit out some Makefiles.  `make`:

```
Making all in lldb
if test "/Users/me/local/bin/emacs" != "no"; then \
      am__dir=. am__subdir_includes=''; \
      case core.elc in */*) \
        am__dir=`echo 'core.elc' | sed 's,/[^/]*$,,'`; \
        am__subdir_includes="-L $am__dir -L ./$am__dir"; \
      esac; \
      test -d "$am__dir" || .././install-sh -c -d "$am__dir" || exit 1; \
      /Users/me/local/bin/emacs --batch \
          \
        $am__subdir_includes -L . -L . \
        --eval "(defun byte-compile-dest-file (f) \"core.elc\")" \
        --eval "(unless (byte-compile-file \"core.el\") (kill-emacs 1))"; \
    else :; fi

In toplevel form:
core.el:16:1:Error: Cannot open load file: no such file or directory, load-relative
make[1]: *** [core.elc] Error 1
make: *** [all-recursive] Error 1
```

Am I on the right track?  Any tips on getting this installed and making use of realgud-lldb in Emacs?  Thanks for the help,

Dan

**Comment by rocky** [comment]
I'm sorry there are problems that you are having here.

 As mentioned in the [wiki](https://github.com/rocky/emacs-dbgr/wiki/How-to-Install#the-full-instructions), emacs-dbgr is dependent on the [load-relative](https://github.com/rocky/emacs-load-relative) package and a couple of other packages. Did you install these into /Users/me/local/share/emacs/site-lisp ? 

When you ran _make_,  the error message shown above says it is not finding that emacs package.  If you have _load-relative_ around, you can make sure emacs running in batch mode finds it by setting _ELCFLAGS_. For example:

```
   ELCFLAGS='-L /usrs/me/.emacs.d/load-relative -L <place-loc-change-is-installed>' make
```

**Comment by DanSchoppe** [comment]
Thanks for the reply, @rocky .  I used ELCFLAGS to point to the installed directories for test-simple, load-relative, loc-change, and realgud. All of these must have gotten installed in their respective ~/.emacs.d/elpa/[package-name] directories when I installed realgud from Melpa.  My make error changed to `core.el:17:1:Error: Cannot open load file: no such file or directory, realgud-track`.  I don't see realgud-track anywhere on my filesystem.  Looking ahead in the lldb/core.el file, I don't see realgud-core or realgud-lang on my filesystem either.

As mentioned on your emacs-dbgr wiki, I do have to load the realgud library in Emacs.  Perhaps I need to do the same for this make file?

Thanks again for the help,
Dan

**Comment by rocky** [comment]
@DanSchoppe I think if you change line 17 of core.el as I've done in  commit  28d43e6 things will be fine, assuming you've installed the bulk code for realgud.

Pulling in realgud should pull in all of those other sub components.

**Comment by DanSchoppe** [comment]
@rocky So I blew away my directory and did a fresh git clone and confirmed I received your latest commit.  I went through the same workflow and used ELCFLAGS to specify Melpa directories of test-simple, load-relative, loc-changes, realgud, and list-utils.  Now I'm getting `init.el:19:1:Error: Cannot open load file: no such file or directory, realgud-regexp`

Thanks for your help,
Dan

**Comment by rocky** [comment]
@DanSchoppe again sorry for this. Commit 90d5673 tries to address this.

But again this is the same kind of problem as you were experiencing before. So let me try to explain what's going on...

I took this code from working inside realgud and moved it outside. 

load-relative is for _internal_ linking and allows those pieces of the code needed to be used more surgically. I do this because I want to be able to run and test each file in isolation with the minimum of additional files that get pulled in. See http://rocky.github.io/NYC-Lisp-Elisp-talk/ for motivation on this. 

However when you are using realgud as an external package, things are easier because that's the more conventional Emacs case, so a simple `(require 'realgud)` is all that is needed. 

Lastly, when this is all worked out, you'd do me a favor if you write this up as a wiki entry in the emacs-dbgr github repository on how to write a new, _external_ debugger outside of the project. You can probably cite this repository (or yours) as an example to look at.

**Comment by DanSchoppe** [comment]
@rocky Hi rocky,

Sorry for the delay.  Thanks for the commit to move this along.  After applying the commit I end up with more errors from lldb.el:16 requiring list-utils.  When I comment that line out I track-mode.el:20 gives me an error for requiring realgud.  Then regexp-helper, test-lldb, and test-regexp-lldb complain.  After commenting out all of these includes I end up with:

```
Making all in lldb
make[1]: Nothing to be done for `all'.
Making all in test
if test "/Users/me/local/bin/emacs" != "no"; then \
      am__dir=. am__subdir_includes=''; \
      case test-regexp-lldb.elc in */*) \
        am__dir=`echo 'test-regexp-lldb.elc' | sed 's,/[^/]*$,,'`; \
        am__subdir_includes="-L $am__dir -L ./$am__dir"; \
      esac; \
      test -d "$am__dir" || .././install-sh -c -d "$am__dir" || exit 1; \
      /Users/me/local/bin/emacs --batch \
          \
        $am__subdir_includes -L . -L . \
        --eval "(defun byte-compile-dest-file (f) \"test-regexp-lldb.elc\")" \
        --eval "(unless (byte-compile-file \"test-regexp-lldb.el\") (kill-emacs 1))"; \
    else :; fi

In toplevel form:
test-regexp-lldb.el:13:24:Warning: global/dynamic var `s1' lacks a prefix
test-regexp-lldb.el:14:24:Warning: global/dynamic var `pos' lacks a prefix
test-regexp-lldb.el:45:29:Warning: reference to free variable
    `realgud:gdb-pat-hash'
test-regexp-lldb.el:44:7:Warning: assignment to free variable `realgud-bt-pat'
test-regexp-lldb.el:54:45:Warning: reference to free variable `realgud-bt-pat'
test-regexp-lldb.el:54:7:Warning: assignment to free variable `realgud-bt-re'
test-regexp-lldb.el:57:31:Warning: reference to free variable `realgud-bt-re'

In end of data:
test-regexp-lldb.el:109:1:Warning: the following functions are not known to be
    defined: test-simple-start, make-realgud-cmdbuf-info,
    realgud-loc-pat-regexp, realgud-loc-pat-file-group,
    realgud-loc-pat-line-group, note, assert-t, cmdbuf-loc-match,
    assert-equal, realgud-cmdbuf-info-file-group,
    realgud-cmdbuf-info-line-group, prompt-match, end-tests
Wrote test-regexp-lldb.elc
make[1]: Nothing to be done for `all-am'.
```

Honestly I don't know what the desired output files are for this build.  Am I rebuilding all of realgud or am I just building an lldb file that will be used by realgud to expand it's functionality?  It looks like the `make install` mentions these files: core.el(c), init.el(c), track-mode.el(c), regexp-helper.el(c), test-lldb.el(c), and test-regexp-lldb.el(c).

Thanks,
Dan

**Comment by rocky** [comment]
@DanSchoppe 

Make is a framework for running a set of canned commands based in dependencies.  Strictly speaking one doesn't need make at all whatever commands it runs you could run without make or run an equivalent command. (I hope and expect you already knew this.)

What "make all" does is byte-compile all of the emacs lisp (el) files into elc files. Strictly speaking, you don't need to do this and just run the lisp code without compiler optimization or you could use byte-recompile-directory from inside emacs. 

"make install" copies the emacs lisp files and compiled files to a place that emacs is supposed to find it when you run M-x "load-library". Again strictly speaking you don't need that if everything is working, especially "load-relative", and you "load-file" lldb.el inside emacs.  

"make test" is also strictly speaking not essential, but I like to think of it as essential. 

> lldb.el:16 requiring list-utils.

A cursory scan doesn't show list-utils is used. I think I changed the code to use realgud:flatten instead. That said, the package realgud uses it, so it needs to be around somewhere on your system. And so this is a sign of the batch el compiler not being able to find it, which might be a problem. Inside emacs you should at least be able to do a "M-x load-library list-utils"  and M-x load-library realgud.

As for: 

```
Making all in test
test-regexp-lldb.el:13:24:Warning: global/dynamic var `s1' lacks a prefix
...
```

I feel that you really should be able to figure this out on your own.  There was a time when I saw a message like that for the first time. I was able to figure it out on my own and hope you can too. 

If not, I am having serious doubts about your being able to complete this project more on your own.

**Comment by DanSchoppe** [comment]
@rocky Thanks for the reply.  I think there was a misunderstanding.  I found your comment in [this thread](https://github.com/rocky/emacs-dbgr/issues/30) about realgud-lldb project and assumed realgud-lldb was in a basic, yet functional state.  That was my mistake.  When I pulled down the source and it wouldn't compile, I figured I had misconfigured something or there was a simple difference on my system that was causing the build to fail, and that's what led me to post here.  It didn't occur to me that this repo was in a nun-functional state.

I was approaching this project as a potential user with a desire to contribute fixes and/or improvements where I was able.  However, I now see that you intended me to make the changes necessary to bring this project to a functional state.  This misunderstanding is why I wasn't digging into compilation errors.  I can understand your frustration with my queries.

I'm truly sorry for this misunderstanding.  Your responses were timely and thorough.  I'm certainly not alone in hoping for a decent lldb experience within Emacs, and I'm hoping someone will be able to pick up on this project.  I'm sure realgud is a great extension and has a great future.  

Thanks again for the help this last week,
Dan

**Comment by rocky** [comment]
From the thread you cited: 

> Sure. Make a stab at it. If I see a serious effort made, I'll help out.

...

> I think it best to close this issue here. Given the facts, I think this is now best done outside of this project.

**Comment by rocky** [comment]
With commit  386588a I think this is now installable and minimally usable.

As it says in [`INSTALL.md,` ](https://github.com/realgud/realgud-lldb/blob/master/INSTALL.md) make sure `realgud` and `test-simple` are available inside emacs. And then from inside emacs run: 

``` lisp
(compile (format "EMACSLOADPATH=:%s:%s ./autogen.sh" (file-name-directory (locate-library "test-simple.elc")) (file-name-directory (locate-library "realgud.elc"))))
```

If this does it, close the issue.

**Comment by DanSchoppe** [comment]
Hi rocky,

Thanks for the update. I was able to install and run realgud-lldb.

The autogen.sh script ran fine and generated makefiles. Instead of using ELCFLAGS as you described above to manually indicate libraries that weren't being automatically found, I forced emacs to load packages in batch mode according to:
https://www.emacswiki.org/emacs/BatchMode#toc8
Specifically, I added this to common.mk.in:

```
--eval "(progn(package-initialize))" \
```

I'm not entirely sure why the required emacs packages (load-relative, etc) weren't automatically located. Perhaps because they're stored in ~/.emacs.d/elpa/ instead of ~/local/share/emacs/site-lisp.  Regardless, I was happy to get it working!

Dan
